### PR TITLE
Add satisfiesOnlyOnce to ObjectEnumerableAssert

### LINKED ISSUE #2675: Add method satisfyOnlyOnce for ListAssert
#### Feature summary
Would it be possible to add method something like `satisfyOnlyOnce` for `ListAssert`. It would work the same way `containsOnlyOnce` but with `satisfy`?

#### Example

```java

// positive case
Comment comment1 = new Comment();
comment.setText("bla_1");

Comment comment2 = new Comment();
comment.setText("bla_2");
assertThat(List.of(comment1, comment2)).satisfyOnlyOnce(comment -> assertThat(comment.getText).isEqualTo("bla_1"));

// negative case 1

assertThat(List.of()).satisfyOnlyOnce(comment -> assertThat(comment.getText).isEqualTo(bla1));

// negative case 2
Comment comment1 = new Comment();
comment.setText("bla_1");

Comment comment2 = new Comment();
comment.setText("bla_1");
assertThat(List.of(comment1, comment2)).satisfyOnlyOnce(comment -> assertThat(comment.getText).isEqualTo("bla_1"));
```

### PR DESCRIPTION
**Pull request by StefanBratanov**
#### Check List:
* Fixes #2675 
* Unit tests : YES
* Javadoc with a code example (on API only) : YES
* PR meets the [contributing guidelines](https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md)

I wasn't sure which version to add for `@since`

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java [hunk 1]
```diff
@@ -3746,6 +3746,73 @@ protected SELF satisfiesExactlyInAnyOrderForProxy(Consumer<? super ELEMENT>[] re
     return myself;
   }
 
+  /**
+   * Verifies that there is exactly one element of the iterable elements that satisfies the {@link Consumer}.
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java:3750]: `of the iterable elements` -> `of the array under test`

Comment (joel-costigliola) [reply src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java:3750]: BTW thanks for having javadoc for the different type where `satisfiesOnlyOnce` was added!

File: src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java [hunk 2]
```diff
@@ -3746,6 +3746,73 @@ protected SELF satisfiesExactlyInAnyOrderForProxy(Consumer<? super ELEMENT>[] re
...
+   *
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java:3751]: add a paragraph to separate the examples from the description, a line like should be enough:
```
  * <p>
```

File: src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java [hunk 3]
```diff
@@ -3746,6 +3746,73 @@ protected SELF satisfiesExactlyInAnyOrderForProxy(Consumer<? super ELEMENT>[] re
...
+   * Examples:
+   * <pre><code class='java'> String[] starWarsCharacterNames = {"Luke", "Leia", "Yoda"};
+   *
+   * // these assertions succeed:
+   *
+   * assertThat(starWarsCharacterNames).satisfiesOnlyOnce(name -&gt; assertThat(name).contains("Y")) // matches only "Yoda"
+   *								   .satisfiesOnlyOnce(name -&gt; assertThat(name).contains("Lu")) // matches only "Luke"
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java:3758]: align with the other `.satisfiesOnlyOnce`

File: src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java [hunk 4]
```diff
@@ -3746,6 +3746,73 @@ protected SELF satisfiesExactlyInAnyOrderForProxy(Consumer<? super ELEMENT>[] re
...
+   * // this assertion fails because no element contains "Han"
+   * assertThat(starWarsCharacterNames).satisfiesOnlyOnce(name -&gt; assertThat(name).contains("Han"));</code></pre>
+   *
+   * @param requirements the {@link ThrowingConsumer} that is expected to be satisfied only once by the elements of the given {@code Iterable}.
+   * @return this assertion object.
+   * @throws NullPointerException if the given requirements are {@code null}.
+   * @throws RuntimeException rethrown as is by the given {@link ThrowingConsumer} or wrapping any {@link Throwable}.    
+   * @throws AssertionError if the requirements are not satisfied only once
+   * @since 3.24.0
+   */
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java:3802]: similar comments as `satisfiesOnlyOnce(Consumer<? super ELEMENT> requirements)`

File: src/main/java/org/assertj/core/api/AtomicReferenceArrayAssert.java [hunk 5]
```diff
@@ -3822,6 +3821,73 @@ protected AtomicReferenceArrayAssert<T> satisfiesExactlyInAnyOrderForProxy(Consu
...
+   * assertThat(starWarsCharacterNames).satisfiesOnlyOnce(name -&gt; assertThat(name).contains("a")); // matches "Leia" and "Yoda"
+   *
+   * // this assertion fails because no element contains "Han"
+   * assertThat(starWarsCharacterNames).satisfiesOnlyOnce(name -&gt; assertThat(name).contains("Han"));</code></pre>
+   *
+   * @param requirements the {@link Consumer} that is expected to be satisfied only once by the elements of the given {@code Iterable}.
+   * @return this assertion object.
+   * @throws NullPointerException if the given requirements are {@code null}.
+   * @throws AssertionError if the requirements are not satisfied only once
+   * @since 3.24.0
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AtomicReferenceArrayAssert.java:3846]: same comments as the ones in `AbstractObjectArrayAssert`

File: src/main/java/org/assertj/core/api/ObjectEnumerableAssert.java [hunk 6]
```diff
@@ -1549,6 +1549,61 @@
...
+  /**
+   * Verifies that there is exactly one element of the iterable elements that satisfies the {@link Consumer}.
+   *
+   * Examples:
+   * <pre><code class='java'> List&lt;String&gt; starWarsCharacterNames = list("Luke", "Leia", "Yoda");
+   *
+   * // these assertions succeed:
+   *
+   * assertThat(starWarsCharacterNames).satisfiesOnlyOnce(name -&gt; assertThat(name).contains("Y")) // matches only "Yoda"
+   *								   .satisfiesOnlyOnce(name -&gt; assertThat(name).contains("Lu")) // matches only "Luke"
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/ObjectEnumerableAssert.java:1561]: align with other `.satisfiesOnlyOnce`

File: src/main/java/org/assertj/core/error/ShouldSatisfyOnlyOnce.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.assertj.core.util.VisibleForTesting;
+
+/**
+ * Creates an error message indicating that an assertion that verifies that requirements are not satisfied only once.
+ */
+public class ShouldSatisfyOnlyOnce extends BasicErrorMessageFactory {
+
+  @VisibleForTesting
+  public static final String REQUIREMENTS_SHOULD_BE_SATISFIED_ONLY_ONCE = "%nExpecting actual:%n  %s%nto satisfy the requirements only once but it did %s times";
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/ShouldSatisfyOnlyOnce.java:23]: I think we can improve the error message:

- reword the beginning, I'm thinking of: `Expecting exactly one element of actual:%n  %s%nto satisfy the requirements but ...`
- a specific message when no elements satisfied the requirements, ex: `but no element did` or `none did`
```
Expecting exactly one element of actual:
  [e1, e2, e3, e4]
to satisfy the requirements but none did
```
- display list the offending elements when multiple elements satisfied the requirements, that will help users understanding the root cause. ex:

```
Expecting exactly one element of actual:
  [e1, e2, e3, e4]
to satisfy the requirements but these 2 elements did: 
  [e1, e3]
```

File: src/main/java/org/assertj/core/internal/Iterables.java [hunk 8]
```diff
@@ -1205,6 +1206,21 @@ public <E> void assertSatisfiesExactlyInAnyOrder(AssertionInfo info, Iterable<?
...
   }
 
+  public <E> void assertSatisfiesOnlyOnce(AssertionInfo info, Iterable<? extends E> actual,
+                                          Consumer<? super E> requirements) {
+    assertNotNull(info, actual);
+    requireNonNull(requirements, "The Consumer<? super E> expressing the requirements must not be null");
+
+    long countOfSatisfactions = stream(actual)
+                                              .filter(byPassingAssertions(requirements))
+                                              .count();
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Iterables.java:1216]: As per my comments on the error message, we want to capture the elements satisfying the requirements to report them if more than one did

File: src/main/java/org/assertj/core/internal/Iterables.java [hunk 9]
```diff
@@ -1387,7 +1403,7 @@ private <E> List<E> notSatisfyingCondition(Iterable<? extends E> actual, Conditi
   }
 
   private <E> List<E> satisfiesCondition(Iterable<? extends E> actual, Condition<? super E> condition) {
-    return stream(actual).filter(o -> condition.matches(o)).collect(toList());
+    return stream(actual).filter(condition::matches).collect(toList());
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Iterables.java:1406]: thanks :)

File: src/test/java/org/assertj/core/internal/iterables/Iterables_assertSatisfiesOnlyOnce_Test.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+import org.assertj.core.internal.IterablesBaseTest;
+import org.assertj.core.test.jdk11.Jdk11.List;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests for <code>{@link Iterables#assertSatisfiesOnlyOnce(AssertionInfo, Iterable, Consumer)}</code>.
+ *
+ * @author Stefan Bratanov
+ */
+class Iterables_assertSatisfiesOnlyOnce_Test extends IterablesBaseTest {
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/internal/iterables/Iterables_assertSatisfiesOnlyOnce_Test.java:38]: use `expectAssertionError` as indicated in https://github.com/assertj/assertj-core/blob/main/CONTRIBUTING.md that will avoid to write `assertThat(error).isInstanceOf(AssertionError.class);`

**Review by joel-costigliola — CHANGES_REQUESTED** [review]
thanks @StefanBratanov PR is overall good :) most of my comments are minor improvements except for the error message one where there is a bit more work.

**Comment by StefanBratanov** [comment]
Thank you for the review. :) Will work on the required changes in the next few days. They all make sense to me.

File: src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java (hunk 3 above)
Comment (StefanBratanov) [reply src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java]: This one is a bit weird, in my IDE it is aligned, i ended up changing it from the GitHub editor. But now it is wrong on my IDE. It is worth having a double check. I use Windows by the way, could be potentially some line endings not sure.

Comment (StefanBratanov) [reply src/main/java/org/assertj/core/api/AbstractObjectArrayAssert.java]: Great! Thanks. I did commit the other changes as requested by the way, so they are ready for review.

**Comment by joel-costigliola** [comment]
Integrated thanks @StefanBratanov! 
FYI the javadoc alignment issue was due either to some non visible characters I believe, deleting them and adding spaces accordingly fixed it.

**Comment by StefanBratanov** [comment]
> Integrated thanks @StefanBratanov! FYI the javadoc alignment issue was due either to some non visible characters I believe, deleting them and adding spaces accordingly fixed it.

Perfect. Thank you! :)

**Pull request closed without merging**